### DISCUSSION TITLE
RFC: keep form values and it's defaults in sync with remote data

### DISCUSSION THREAD

**Discussion by Moshyfawn** [react-hook-form/react-hook-form#8233] [RFC]
# Intro

This is a big topic about managing a close-to-real-time _(in some cases, actually real-time)_ form data. I think it's important to introduce a solid way to **sync** the remote data with the currently active _(editable)_ form.

# The current state of things

At this time, React Hook Form is focused on "traditional" more Redux-like type of data flow _(a very stretch comparison, IK. Gotta anchor to smth lol)_: you fetch some `async` default values and `reset` the form when you get the data, not thinking about the remote data anymore; you've set the client-side form values and operate the client state, assuming the remote data never changes without the current form submission _(mutating remote data on form submission)_
```js
const { reset } = useForm()
useEffect(() => {
  if (remoteData) {
    reset(remoteData)
  }
}, [reset, remoteData])
```

# The problem

The issue arises when the data becomes more "unstable": introduction of background re-fetches, stale data, `subscriptions`, etc.

The JS eco-system is greatly advancing in the field of cache management, back-end - front-end data `sync`ing. Just look at packages like [SWR](https://github.com/vercel/swr) and [Apollo](https://github.com/apollographql/apollo-client). Even "the one store - 99 problems" Redux is moving towards dividing the data cache and the client store: [RTK Query](https://github.com/reduxjs/redux-toolkit).

I'm going to use [the React Query cache management _(and a ton more)_ package](https://github.com/tannerlinsley/react-query) as a base to showcase the ever-changing data use-case for simplicity-sake.

Consider some back-end data entry you can update _(mutate)_ via your form on the client _(your basic CRUD, if you will)_:
```json
{
  "name": "John Doe"
}
```
The potential form setup to update the entry with React Hook Form can look like this:
```js
const { register } = useForm()

return (
  <form>
    <input {...register("name")} defaultValue="" />
  </form>
)
```
Now, our data is stored and the back-end and we need to retrieve it to initialize our form with the actual default values with out entry data. We don't have the needed data to be set as form's `defaultValues` initially. As I've mentioned previously, the current React Hook Form recommended pattern is to `reset` the form's values as soon as our `async` data comes in. Let's try applying it to our form _(Keep in mind that we're not looking at such concepts as `suspense` as it's a whole another topic)_
```diff
// INFO: data can be `{ name: string }` or `undefined`
+ const { data: user } = useQuery(["users", { userId }], getUser)
  const {
    register,
+   reset
  } = useForm()

+ useEffect(() => {
+   if (user) {
+     reset(user)
+   }
+ }, [user, reset])

  return (
    <form>
      <input
        {...register("name")}
-       defaultValue=""
      />
    </form>
  )
```
Provided our back-end user data changes only when we submit the form successfully, problem solved.

Introducing background re-fetch. I won't go into great detail about how it works or what it means for our form. For that you can read up the @tkdodo's blog-post [React Query and Forms](https://tkdodo.eu/blog/react-query-and-forms). For our form, background refetch means that our client state form data will be eventually `synced` with the remote data during the life-time of our form, meaning, with our current setup the user form input will be `reset` with the back-end data, which can potentially differ; not a greate UX at all.

You can check out [this codesanbox which demonstrates our current issue](https://codesandbox.io/s/rough-platform-v6xuzd?file=%2Fsrc%2FForm.tsx):
1. Open the user form
2. Enter some value different from defaults
3. Cause background refetch
   - Un-focus the current browser window _(or switch to a different tab)_
   - Click Re-fetch User button _(not so much of a background re-fetch, but a re-fetch nonetheless)_
4. See the form input's data has changed

# The current solution

There's a lot going on and there're multiple moving pieces to this problem. There's no one way to tackle this topic.

It's tough to solve an issue, the solution to which depends on the business requirements, app architecture, performance, etc. etc. Simply put: it's case-by-case study. To list a few:
- Do you want to override the user's form input with the new remote data?
- Do you want to show a notification about the new remote data?
- Give the user control over `sync`ing client and server state
- Ignore the background refetch all together?

That said, there're a couple of things you can decide on the form's side.

The solution that comes to mind is to check the `dirtyFields` values and override only the yet to be modified fields, leaving the dirty fields with the current user-entered values. This would look something like
```js
const { data: user } = useQuery(["users", { userId }], getUser)

const { formState, getValues, reset } = useForm()
const { dirtyFields } = formState

useEffect(() => {
  if (user) {
    const newDefaultValues = Object.fromEntries(
      Object.entries(user).filter(([userKey]) => !dirtyFields[userKey])
    )
    const currentValues = getValues()

    reset(
      {
        ...currentValues,
        ...newDefaultValues
      },
      {
        keepDirty: true
      }
    )
  }
}, [user, dirtyFields, getValues, reset])
```
[Here's a codesanbox to play around with it live](https://codesandbox.io/s/sad-pine-9qph99?file=%2Fsrc%2FForm.tsx)

This approach while being a bit convoluted, allows to us to override only the currently unmodified form fields, leaving the user's progress alone. Is it the solution? No, it leaves a lot to be desired in terms of UX: updating form values, even the user hasn't yet touched them can easily get confusing.

@tkdodo provides a slightly different approach to managing the form values updates using the controlled inputs. You can see [his example here](https://codesandbox.io/s/react-hook-form-submit-only-dirty-fields-ol5d2).

The proper solution would combine something from the method mentioned above with potentially a better React Hook Form API to update the intended fields + the user notification system + potentially a completely different approach + ...

You can notice how untapped this question is. Hence we need your help :P

# Outro

I'm very curious to hear your opinion on the matter. Please, share you experience and potential solutions you might have came up with, so we can all discuss it and improve React Hook Form and sequentially the developer's quality of life _(your life, lol)_

_P.S. For those who are not familiar with the cache management concept described above, please, take a look at the greatest information source there's about React Query and back-end data synchronization in general: @tkdodo's [blog](https://tkdodo.eu/blog). Or feel free to ask me questions on the topic here._

_P.P.S. Here's a list of discussion resources:_
- _@tkdodo's [RQ-forms blog comments](https://github.com/TkDodo/blog-comments/discussions/79)_
- _@ciaoimxiao's ["`keepDirtyOnReinitialize` flag" Twitter thread](https://twitter.com/moshyfawn/status/1517416661648818176)_
- _@tkdodo's [RQ-forms Twitter thread](https://twitter.com/moshyfawn/status/1513152744613371914)_

_P.P.P.S. Sry for a long and crumpled description. it's such a big topic; my little brain doesn't know how to bite into it at times xD_

**Comment by bluebill1049** [comment]
relevant: https://github.com/react-hook-form/react-hook-form/issues/2335

**Reply by Moshyfawn** [reply]
Yep!

Tho, the initial default `async` values _(the topic raised in #2335)_ has it's own context to discuss. There's a `suspense` solution also.

The OG post is more about the whole life-cycle of the form

**Comment by TkDodo** [comment]
I have to admit I haven't fully understood the `reset` api 😅 , but generally, I don't think it's a bad idea to have this done with a useEffect. Users already have great control over what they want to do:

- overwrite the user input with every background data change? Just call `reset` like described:
```
useEffect(() => {
  if (user) {
    reset(user)
  }
}, [user, reset])
```
- show a toast with action when new data comes in to let the user choose what to do? Easy:

```
useEffect(() => {
  if (user) {
    // some async dialog that returns a Promise
    dialog.confirm().then(() => reset(user))
  }
}, [user, reset])
```

However, what is missing is that you can keep your own input on big forms, but get untouched fields updated by background updates. What I would like to see is a way to sync only the `defaultValues`:

```
const {
   register,
   reset,
   defaultValue: { name: "" }
} = useForm()

useEffect(() => {
  if (user) {
    // hypothetical api that only updates the defaultValues
    reset({ defaultValues: user })
  }
}, [user, reset])
```

now if we have changed the name already, this wouldn't do anything. Maybe, if we have changed the name to `Alice`, and the background update comes in that also sets the name to `Alice`, our field would be reset to be in a non-dirty state, because it is now the same as the default value. Maybe the value itself could even be reset to `undefined`. That would be sweet.

If we have set the name to `Bob` in the meantime, and the background update yields `Alice`, the `defaultValue` would be `Alice`, but the value would still be `Bob`, meaning our field is still dirty,. But if we now reset, it would go back to `Alice` instead of empty string.

I hope I'm making somewhat sense 😅

**Reply by bluebill1049** [reply]
Thanks, @TkDodo for the awesome feedback and input!

I quite like this idea (API design): `reset({ defaultValues: user })` which translates to the current API is something like 

`reset(undefined, { defaultValues: users })` as the first argument would update both both values and defautlValues (to keep form data sync)

**Comment by bluebill1049** [comment]
```tsx
reset(data, {
  keepDirtyValues: true
})
```

I think this API is probably closest to the existing pattern. options to have a new `keepDirtyValues `, so `defaultValues` will be updated, and `values` will be updated except for those dirtied fields.

**Reply by bluebill1049** [reply]
Also, I think it's fine to let `reset` be associated with `useEffect` for now, it does give user the fine control on how and when to `reset`, rather keep more logic within the library. After all, my personal preference is still to keep this library small and as primitive as it can be.

**Reply by bluebill1049** [reply]
here is  a working example: https://codesandbox.io/s/tender-panini-o8to91?file=/src/App.js

**Reply by TkDodo** [reply]
> Also, I think it's fine to let reset be associated with useEffect

yes i think so too 👍

Regarding the example: the difference is that it does set the value and not the defaultValue. So it is treated as if the user had made that input, which is not true. A button "undo my changes" that would just call reset() would not work because it would also undo the async reset from the effect, which we would like to keep.

**Reply by bluebill1049** [reply]
Good pick, this should be the expected behaviour then: https://codesandbox.io/s/tender-panini-o8to91?file=/src/App.js

**Reply by TkDodo** [reply]
hmm no, that't not what I'd expect to happen. let me elaborate with some examples:

- default values are empty
- async data comes in with `bill, luo`
- it is filled to default values
- if I know change `bill` to `bob`, then click reset, it goes back to `bill, luo`
- that works ✅ 

---

- default values are empty
- I change `firstName` to `bob`
- async data comes in with `bill, luo`
- on the screen, we see `bob, luo` because my local change takes precedence
- if I know click reset, it goes back to `bill, luo`
- that doesn't work at the moment ❌

**Reply by bluebill1049** [reply]
> on the screen, we see bob, luo because my local change takes precedence
> if I know click reset, it goes back to bill, luo

That's an interesting UX question, why would it revert to "Bill, Luo", It feels a bit confusing to me as I (the "user") have overwritten the data locally. The server data should only overwrite data that doesn't "exist" or has not been modified.

1. user changes the input value
2. server data update the unmodified value
3. user click reset again 
4. reset the form back to the last user modified form state 
6. if we revert back to the latest server state, would that confuse the user I have modified that field but I never entered such value during this session

I try to illustrate in the following video clip, I may be missing understanding the concept here.

https://user-images.githubusercontent.com/10513364/164884815-2f2c28e1-3ece-484f-b88c-93ab3219c3b4.mov

**Reply by TkDodo** [reply]
maybe I'm misunderstanding what `reset` should be doing. I thought it should reset the form to it's initial state, which is its default values. Keeping background refetches out of equation for a second: If I initialize a form with:

```
useForm({
    defaultValues: {
      firstName: "hello",
      lastName: "world"
    }
  })
```

and then I change the first name and click reset, it resets back to `firstName: hello, lastName: world`. My take is that if a background refetch comes in, it would just update those default values. Because semantically, I have a user entity in the database, and I want to initialize my form with the values I have in the database. if the database updates and I get this data, e.g. due to a background refetch or because of a push by a server-side event, I would like the default values to "become" these values. This has two advantages over just setting the "values" of the form (which is what the first parameter of `reset` is doing):
- only fields where I haven't made changes show the defaultValues. So fields that are dirty will not be overwritten by background updates, which is good. Never overwrite user input without asking the user first :)
- when I do `reset()`, it goes back to the latest server state (= the default values). Again, we're not interested here in resetting to the last user input, but we'd basically like that button to behave like a page refresh: Discard all my changes, and show me the latest values I have in the database.

Another good api would be e.g.:

```
const { data} = useQuery(['user'], () => fetchUser())
useForm({
    defaultValues: {
      firstName: data?.firstName,
      lastName: data?.lastName
    }
  })
```

and `useForm` would just update the `defaultValues` when they change. But that might not be what developers expect, because:
- `const [value, setValue] = useState(initialValue)` also doesn't keep track of when `initialValue` changes
- `<input defaultValue={myVariable} />` also doesn't keep track of when `myVariable` changes

so I think the imperative `reset` would be fine:

```
const { data} = useQuery(['user'], () => fetchUser())
const { setDefaultValues } = useForm({
    defaultValues: {
      firstName: data?.firstName,
      lastName: data?.lastName
    }
  })

useEffect() => {
  if (data) {
    setDefaultValues(data)
  }
}, [data])
```

**Reply by Moshyfawn** [reply]
The `reset` API is in control of both **form values** and the **form's state** reset. There's a list of `reset` options to determine what part of the form gets reset (you can checkout the `reset` options description [in the RHF `reset` docs](https://react-hook-form.com/api/useform/reset)):
- keepErrors
- keepDirty
- keepValues
- keepDefaultValues
- keepIsSubmitted
- keepTouched
- keepIsValid
- keepSubmitCount

So, it's not only the form values reset _(like the native Web `reset` method would be)_

The current RHF version already contains the API to reset the `defaultValues` **only**; the `defaultValues` get reset by default, unless the `keepDefaultValues` `reset` option provided `keepDefaultValues: true`. You can keep the current form values _(modified or not)_ by providing the `keepValues` `reset` option. It doesn't check whether the field is dirty or not; all current form values are left as is.

If we'd apply the concept to our question at hand: to deal with background re-fetches, we can have a side-effect to set the `defaultValues` when the data "get there" along with the form values _(like actual default / initial form values)_
```js
const { user, refetchUser } = useUser({ userId })
// INFO: a custom hook that returns the previously rendered value
const prevUser = usePrevious(user)

useEffect(() => {
  if (!prevUser && user) {
    reset(user)
  }
}, [user, prevUser, reset])
``` 
Later we can have a side-effect to update the` defaultValues` only to get the new form values defaults to compare the field's "dirtiness" to.
```js
const { user, refetchUser } = useUser({ userId })

const prevUser = usePrevious(user)

useEffect(() => {
  // INFO: we want to ensure that we already have the initial remote defaultValues set (prevUser != null)
  if (prevUser && user) {
    reset(user, {
      keepValues: true
    })
  }
}, [user, prevUser, reset])
```
 
[Here's an example codesandbox](https://codesandbox.io/s/immutable-http-tl059t?file=%2Fsrc%2FForm.tsx):
1. Open the form
2. Click Refetch User button to update the "new" defaultValues
3. See the new defualt `age` value
4. Set the `age` field's value to the new default age value
5. See that the `age` field is considered non-dirty due to comparing it's current value to it's new default one

https://user-images.githubusercontent.com/16290753/164911783-374c7c54-e446-4199-a76d-90c25975bdfa.mp4

Speaking of `keepValues`, here's an idea: instead of `keepValues: boolean`, what if it would take, say, a literal `string` like
```ts
keepValues: "all" | "unmodified" | "dirty"
```

The example `keepValues` values are provided from the top if my head. It might not be feasible to have the `dirty` option, for example. But I hope you get the point.

**Reply by TkDodo** [reply]
your example shows the issue very well: I refetch the user, and the default values update (internally), but they are not visible on the screen, even though no field is dirty. I would expect non-dirty fields to update visually on the screen as well - or is that a weird ask 😅 ?

also, side node: there is no good way to implement `usePrevious` without technically violating the rules of react. The obvious version reads from a ref during render, which is "not allowed" and can get you in troubles with concurrent features. see: https://twitter.com/brian_d_vaughn/status/1317536300845666304

**Reply by Moshyfawn** [reply]
> I would expect non-dirty fields to update visually on the screen as well - or is that a weird ask 😅 ?

The point of the example below is more to demonstrate the behavior of the `keepValues: true` `reset` option set to this point:
> My take is that if a background refetch comes in, it would just update those default values

I believe, I see where the disconnect between what you expect to happen to `defaultValues` and what RHF is doing is _(even tho, I can't quite put it into words yet lol)_.

Take a look at #8237. It's a `reset` API proposal to handle resetting the specific field values; not resetting the dirty fields.

> there is no good way to implement usePrevious without technically violating the rules of react.

Oh, shoot! That I completely skipped.. Hm, it's supper weird to me that the use of functional components has such a drawback! I guess I can use something like [immer](https://github.com/immerjs/immer) to check the previous value, but that's.. that's interesting! 🤔

**Reply by Moshyfawn** [reply]
> also, side node: there is no good way to implement `usePrevious` without technically violating the rules of react. The obvious version reads from a ref during render, which is "not allowed" and can get you in troubles with concurrent features. see: https://twitter.com/brian_d_vaughn/status/1317536300845666304

I see the most basic `usePrevious` hook implementation [on the official React docs page](https://reactjs.org/docs/hooks-faq.html#how-to-get-the-previous-props-or-state). It uses refs to read the value during render. Isn't it weird?

**Reply by TkDodo** [reply]
>  Isn't it weird?

very! rules of react are "weird" 😅

**Reply by TkDodo** [reply]
https://twitter.com/TkDodo/status/1518301353231138816

**Comment by Moshyfawn** [comment]
I guess the solution with the current API is this weird setup _(reusing the logic from the initial "solution" example from the OG post)_

```js
useEffect(() => {
  if (user) {
    const newDefaultValues = Object.fromEntries(
      Object.entries(user).filter(([userKey]) => !dirtyFields[userKey])
    )
    const currentValues = getValues()

    reset(user, {
      keepValues: true
    })

    reset(
      {
        ...currentValues,
        ...newDefaultValues
      },
      {
        keepDefaultValues: true
      }
    )
  }
}, [user, dirtyFields, getValues, reset])
```

It seems to work as expected. I know it's not pretty, but that's what I can come up with at this time.

[Here's a codesanbox](https://codesandbox.io/s/hardcore-night-u8y26t?file=%2Fsrc%2FForm.tsx%3A680-1103)

https://user-images.githubusercontent.com/16290753/164913572-b09f8f13-1ae5-4219-a376-dbca5c578c09.mp4

**Reply by TkDodo** [reply]
hmm, I added a log statement and there seems to be an infinite render loop somewhere in that code. what I also don't understand is if the fields are still being marked as `dirty` if we call `refetchUser` and then `reset` the form. That's what I wanted to log when I saw the infinite loop.

Anyways, I'm not actually trying to find a workaround on how to make that work currently in user land. I'm more interested in finding an idiomatic api exposed from RHF to make that possible, if you think that's a good idea. If not, I'll likely just stick to control fields like I've shown in my blog :)

**Reply by Moshyfawn** [reply]
Ooops, spotted the infinite loop!

The `dirtyFields` get recalculated on the next render, so they change is not reflected straight away.

I'm not trying to provide a solution with the current API, as it's clearly not there yet. It's more of a middle point to sum up what's the current state of things.

I'm interested in providing an idiomatic solution to the problem as it's clearly a necessary issue to be tackled.

Further progress is #8237 which we gonna look into once @bluebill1049 gets here :P

**Reply by bluebill1049** [reply]
Thanks very much for everyone's input. I think we are heading to a great result here. I have thought about this issue for a while in bed last night. 

There are always going to be two-piece of form values state to be managed. Let's call it 

- **defaultValues**
- **currentFormValues**


1. `keepValues: true`

Update **defaultValues** and skip update all **currentFormValues**, which mark `dirtyFields` and `isDirty`

2. `keepDefaultValues: true`

Update **currentFormValues** and skip update all **defaultValues**, which mark `dirtyFields` and `isDirty`

So above is basically **defaultValues** vs **currentFormValues**

--------------------------------------------------------------------------------------------------------------

3. `keepDirtyValues: true`

Update **defaultValues** and skip all modified fields **currentFormValues**, which mark `dirtyFields` and `isDirty`

I did aim to retain a copy of the user's modified data, but I think that's probably not a valid case for the above which @TkDodo pointed out, we should strictly follow what's involved with `defaultValues` vs `currentFormValues` update by the reset API instead user's action.

https://codesandbox.io/s/tender-panini-o8to91?file=/src/App.js

https://user-images.githubusercontent.com/10513364/164948764-00264e02-b16a-4a80-b115-eb5e61fcb6ba.mov

--------------------------------------------------------------------------------------------------------------

In terms of the API design, I think in the next major version we can improve upon it (breaking change)

```tsx
reset(data); // covers most of user cases
reset(data, { keep: ['values' ] });
reset(data, { keep: ['defaultValues' ] });
reset(data, { keep: ['dirtyValues', 'isDirty', 'errors' ] }); // when user want to keep certain states.
```

while most users' `reset` actions wouldn't be involved with `defaultValues`, it comes in handy when you want to sync with external remote data.

**Reply by mikestopcontinues** [reply]
@bluebill1049 I would avoid the `keep` array syntax. It's non-standard, so it's less intuitive to newbs and non-TS users. Also, it makes conditionals much uglier:

```ts
const opts = {
  keep: [
    isSomeState ? 'dirtyValues' : undefined, // will the type even support undef/null?
  ],
};

// VS...

const opts = {
  keepDefaultValues: isSomeState,
};
```

**Reply by bluebill1049** [reply]
thanks for the feedback! Make sense 👍

**Reply by Moshyfawn** [reply]
> @bluebill1049 I would avoid the `keep` array syntax. It's non-standard, so it's less intuitive to newbs and non-TS users. Also, it makes conditionals much uglier:
> 
> ```ts
> const opts = {
>   keep: [
>     isSomeState ? 'dirtyValues' : undefined, // will the type even support undef/null?
>   ],
> };
> 
> // VS...
> 
> const opts = {
>   keepDefaultValues: isSomeState,
> };
> ```

I see your point. It definitely makes sense, especially the conditional array part.

What I was trying to diviate from is an obnoxious amount of `boolean` options which, say, nooby wouldn't know how they're dependant on each other.

I agree that the `array` syntax doesn't suffices here, but I'd love to explore possible solutions to help with the amount of options we have

**Reply by mikestopcontinues** [reply]
There are quite a lot of options. I wonder if it's possible to reduce them down. E.G.

- keepDirty = keepDirty + keepDirtyValues
- keepSubmit = keepSubmitCount + keepIsSubmitted
- KeepErrors = keepErrors + keepIsValid

**Comment by uncvrd** [comment]
Hi @Moshyfawn ! Out of curiosity, have you discovered a useful way of handling server state with `keepDirtyValues` when using `useFieldArray`? I've been following the react-query forms post from TkDoDo that you have commented on and have had some issues figuring out the best way to implement.

For example, say that your server stores a collection of 2 items to display within your form (to do this we'll utilize `useFieldArray`, while a user edits a form, they decide to remove one of the items within the field array, then react-query refetches data from the server. How would RHF know to not merge in the removed (on the client side) field value? 

Here's a codesandbox which shows this issue: https://codesandbox.io/s/react-keepdirtyvalues-forked-jyezhw?file=/src/App.tsx

Ideally my example would load in data and `reset` within a `useEffect` but since I'm loading in fake data, we're triggering it by a button click, but to recreate:

1. Click `Start Fetch` to "load" in data from the database
2. Change the input value in the first input
3. Remove the second item by clicking `Remove` on the right side
4. Click `Start Fetch` to simulate react-query fetching data again
5. Observe that the first row stays the same with the new modified value (yay! curtesy of the new `keepDirtyValues`) but the previously removed user value (the second row), is obviously added back again.

Any thoughts on the best approach for handling this scenario?

**Comment by dbousamra** [comment]
Hi all. Reading this thread reminds me a lot about our autosave implementation. We allow editing of forms, where the server can occasionally compute new values based on what the user submitted, and we want to update the form with those new values, but not erase any unsynced changes.

Here's our implementation. I know for sure it's not very good, and has a small bug in it (have yet to solve it), but it works for the most part. https://gist.github.com/dbousamra/37d67e2f7b43261d424e7b0d75032d5b

There's some extra stuff in here we do around optimistic locking (taking ownership), that can be ignored.

**Comment by tjhrulz** [comment]
Man, I'm used to random usernames with two names followed by four numbers trying to add me and spam me on every service that isn't GitHub. Now I have to deal with them on GitHub?

For anyone else that shows up there isn't aware the previous comment is spam probably trying to get you to go to some shady website to call you out of money that you'll give them probably via gift cards. Don't click on the links in the comment mentioning a bunch of usernames above and I already reported it so if you would report it to that may speed things up to get it removed.

Edit: well that was quick. Comment is already deleted. Everything on this thread is probably safe now.

**Reply by Moshyfawn** [reply]
I'm trying to keep up with this shit popping up everywhere in real-time. It's being going for almost an hour now: every 5 mins: 2 - 3 spam comments like you just saw. Freaking hell

**Reply by bluebill1049** [reply]
I am getting those as well... annoying

**Reply by Moshyfawn** [reply]
But thx for looking after the community and educating ppl about spam! Appreciate it!

**Reply by Moshyfawn** [reply]
It's also happening in a bunch of other repos of other orgs :(
